File: src/main/java/com/intellij/xtextLanguage/xtext/generator/BnfGenerator.java
package com.intellij.xtextLanguage.xtext.generator;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.PrintWriter;

public class BnfGenerator {
    String extention;
    String myGenDir = "src/main/java/com/intellij/xtextLanguage/xtext/grammar/";
    XtextFileModel myFileModel;
    File myBnfFile;
    PrintWriter out;


    public BnfGenerator(String extention, XtextFileModel fileModel) {
        this.extention = extention;
        myFileModel = fileModel;
        myBnfFile = new File(myGenDir + extention + ".bnf");
        try {
            myBnfFile.createNewFile();
        } catch (IOException e) {
            e.printStackTrace();
        }

    }

    public void generate() throws IOException {
        out = new PrintWriter(new FileOutputStream(myBnfFile));


        generateTokens();
        generateAttributes();
        generateRules();
        generateReferences();

        out.close();
    }

    private void generateReferences() {
        if (myFileModel.myReferences.size() == 0) return;
        else {
            for (ReferenceElement element : myFileModel.myReferences) {
                out.print(element.name);
                out.print("::=");
                out.print(element.referenceType + "\n");
            }
        }
    }

    private void generateRules() {
        for (ParserRule rule : myFileModel.getMyParserRules()) {
            out.print(rule.name);
            out.print(" ::= ");
            out.print(rule.alternatives + "\n");

        }
    }

    private void generateAttributes() {
        out.print(
                "  parserClass=\"com.intellij.xtextLanguage.xtext.parser.XtextParser\"\n" +
                        "\n" +
                        "  extends=\"com.intellij.xtextLanguage.xtext.psi.impl.XtextPsiCompositeElementImpl\"\n" +
                        "  psiClassPrefix=\"Xtext\"\n" +
                        "  psiImplClassSuffix=\"Impl\"\n" +
                        "  psiPackage=\"com.intellij.xtextLanguage.xtext.psi\"\n" +
                        "  psiImplPackage=\"com.intellij.xtextLanguage.xtext.impl\"\n" +
                        "\n" +
                        "  elementTypeHolderClass=\"com.intellij.xtextLanguage.xtext.psi.XtextTypes\"\n" +
                        "  elementTypeClass=\"com.intellij.xtextLanguage.xtext.psi.XtextElementType\"\n" +
                        "  tokenTypeClass=\"com.intellij.xtextLanguage.xtext.psi.XtextTokenType\"\n" +
                        "  psiImplUtilClass=\"com.intellij.xtextLanguage.xtext.psi.impl.XtextPsiImplUtil\"\n" +
                        "  parserUtilClass= \"com.intellij.xtextLanguage.xtext.parserUtilBase.GeneratedParserUtilBaseCopy\"\n" +
                        "  generateTokenAccessors=true\n" +
                        "  generateTokens=true\n" +
                        "\n" +
                        "}");
    }

    private void generateTokens() {
        out.print("{\n" +
                "    tokens = [\n");
        for (TerminalRule rule : myFileModel.getMyTerminalRules()) {
            out.print(rule.name);
            out.print(" = \"regexp:");
            out.print(rule.regexp);
            out.print("\"\n");

        }

        out.print("    ]\n");


    }


}

File: src/main/java/com/intellij/xtextLanguage/xtext/generator/generators/BnfGenerator.kt
package com.intellij.xtextLanguage.xtext.generator.generators

import com.intellij.xtextLanguage.xtext.generator.models.MetaContext
import com.intellij.xtextLanguage.xtext.generator.models.elements.tree.TreeRoot
import java.io.FileOutputStream
import java.io.PrintWriter

class BnfGenerator(extension: String, val context: MetaContext) : AbstractGenerator(extension) {
    fun generateBnf() {
        val file = MainGenerator.createFile(extension.capitalize() + ".bnf", myGenDir + "/grammar")
        val out = PrintWriter(FileOutputStream(file))
        out.print("{\n")
        generateTerminalRules(out)
        generateAttributes(out)
        out.print("}\n")
        generateRules(out)
        generateEnumRules(out)
        out.close()
    }

    fun generateTerminalRules(out: PrintWriter) {
        out.print("    tokens = [\n")
        context.terminalRules.forEach {
            out.print(it.name.toUpperCase() + "=\"regexp:")
            it.alternativeElements.forEach {
                out.print(it.getBnfName())
            }
            out.print("\"\n")

        }
        generateKeywordTokens(out)
        out.print("    ]\n")
    }

    private fun generateKeywordTokens(out: PrintWriter) {
        context.keywordModel.keywords.forEach { out.print("      ${it.name} = \'${it.keyword}\'\n") }
    }

    private fun generateAttributes(out: PrintWriter) {
        out.print("""
          |    parserClass="$packageDir.parser.${extension.capitalize()}Parser"
        
          |    extends="$packageDir.psi.impl.${extension.capitalize()}PsiCompositeElementImpl"
          |    psiClassPrefix="${extension.capitalize()}"
          |    psiImplClassSuffix="Impl"
          |    psiPackage="$packageDir.psi"
          |    psiImplPackage="$packageDir.impl"
          |    elementTypeHolderClass="$packageDir.psi.${extension.capitalize()}Types"
          |    elementTypeClass="$packageDir.psi.${extension.capitalize()}ElementType"
          |    tokenTypeClass="$packageDir.psi.${extension.capitalize()}TokenType"
          |    parserUtilClass= "com.intellij.languageUtil.parserUtilBase.GeneratedParserUtilBaseCopy"
          |    psiImplUtilClass="$packageDir.psi.impl.${extension.capitalize()}PsiImplUtil"
          |    generateTokenAccessors=true
          |    generateTokens=true
          |    extraRoot(".*")= true
                """.trimMargin("|"))
        out.print("\n")
    }


    private fun generateEnumRules(out: PrintWriter) {
        context.enumRules.forEach {
            val rule = it
            out.print("${it.name} ::= ")
            it.alternativeElements.forEach {
                out.print(it.getBnfName())
                if (it != rule.alternativeElements.last()) {
                    out.print(" | ")
                }
            }
            out.print("\n")

        }

    }

    private fun generateRules(out: PrintWriter) {
        out.print("${extension.capitalize()}File ::= ${context.parserRules[0].name}\n")
        context.parserRules.forEach {
            if (context.hasPrivateDuplicate(it)) {
                generatePrivateDuplicate(it, out)
            }
            if (it.isFragment) out.print("private ")
            out.print("${it.name} ::= ${it.getBnfString()}\n")
            val bnfExtensions = mutableListOf<String>()
            if (context.isReferencedRule(it)) {
                bnfExtensions.add("""
                mixin="$packageDir.psi.impl.${extension.capitalize()}NamedElementImpl"
                implements="com.intellij.psi.PsiNameIdentifierOwner"
                methods=[ getName setName getNameIdentifier ]
                """.trimIndent())
            }
            it.superRuleName?.let {
                bnfExtensions.add("extends=$it")
            }
            if (bnfExtensions.isNotEmpty()) {
                out.println("{")
                bnfExtensions.forEach {
                    out.println(it)
                }
                out.println("}")
            }
            out.print("\n")

        }
    }


    private fun generatePrivateDuplicate(rule: TreeRoot, out: PrintWriter) {
        val newName = rule.name + "Private"
        val body = rule.children.map { it.getBnfString() }.joinToString(separator = " ")
        out.println("private $newName ::= $body\n")
    }
}